src/utils/components/TolerationsModal/TolerationsModal.tsx

```diff
@@ -0,0 +1,120 @@
+import * as React from 'react';
+import produce from 'immer';
+
+import { ensurePath } from '@catalog/utils/WizardVMContext';
+import { NodeModel } from '@kubevirt-ui/kubevirt-api/console';
+import { IoK8sApiCoreV1Node } from '@kubevirt-ui/kubevirt-api/kubernetes/models';
+import { K8sIoApiCoreV1Toleration, V1VirtualMachine } from '@kubevirt-ui/kubevirt-api/kubevirt';
+import LabelsList from '@kubevirt-utils/components/NodeSelectorModal/components/LabelList';
+import NodeCheckerAlert from '@kubevirt-utils/components/NodeSelectorModal/components/NodeCheckerAlert';
+import { useIDEntities } from '@kubevirt-utils/components/NodeSelectorModal/hooks/useIDEntities';
+import TabModal from '@kubevirt-utils/components/TabModal/TabModal';
+import { useKubevirtTranslation } from '@kubevirt-utils/hooks/useKubevirtTranslation';
+import { getTolerations } from '@kubevirt-utils/resources/vm';
+import { Operator } from '@openshift-console/dynamic-plugin-sdk';
+import { Form, ModalVariant } from '@patternfly/react-core';
+
+import { TolerationLabel, TOLERATIONS_EFFECTS } from './utils/constants';
+import { getNodeTaintQualifier } from './utils/helpers';
+import TolerationEditRow from './TolerationEditRow';
+import TolerationListHeaders from './TolerationListHeaders';
+import TolerationModalDescriptionText from './TolerationModalDescriptionText';
+
+type TolerationsModalProps = {
+  vm?: V1VirtualMachine;
+  nodes?: IoK8sApiCoreV1Node[];
+  nodesLoaded?: boolean;
+  isOpen: boolean;
+  onClose: () => void;
+  onSubmit: (updatedVM: V1VirtualMachine) => Promise<V1VirtualMachine | void>;
+};
+
+const TolerationsModal: React.FC<TolerationsModalProps> = ({
+  vm,
+  nodes,
+  nodesLoaded,
+  isOpen,
+  onClose,
+  onSubmit,
+}) => {
+  const { t } = useKubevirtTranslation();
+  const {
+    entities: tolerationsLabels,
+    onEntityAdd: onTolerationAdd,
+    onEntityChange: onTolerationChange,
+    onEntityDelete: onTolerationDelete,
+  } = useIDEntities<TolerationLabel>(
+    (getTolerations(vm) || []).map((toleration, id) => ({ ...toleration, id })),
+  );
+
+  const tolerationLabelsEmpty = tolerationsLabels?.length === 0;
+
+  const qualifiedNodes = getNodeTaintQualifier(nodes, nodesLoaded, tolerationsLabels);
+
+  const onSelectorLabelAdd = () =>
+    onTolerationAdd({ id: null, key: '', value: '', effect: TOLERATIONS_EFFECTS[0] });
+
+  const updatedVirtualMachine = React.useMemo(() => {
+    const updatedVM = produce<V1VirtualMachine>(vm, (vmDraft: V1VirtualMachine) => {
+      ensurePath(vmDraft, ['spec.template.template.spec.tolerations']);
+      if (!vmDraft.spec.template.spec.tolerations) {
```

Comment:
add .? will still return undefined but without a crash

Reply:
actuall is it needed? in line 71 can be vmDraft.spec.template.spec.tolerations = updatedTolerations || []

Reply:
added || [] to line 61 where I map the labels, this should have the same affect as suggested :)